xdsl/context.py

```diff
@@ -166,6 +186,43 @@ def get_op(
             return op_type
         raise Exception(f"Operation {name} is not registered")
 
+    def get_optional_type(self, name: str) -> type[Attribute] | None:
+        """
+        Get a type definition from its name if it exists.
+        If the type is not registered, return None unless unregistered types
+        are allowed in the context, in which case return an UnregisteredAttr.
+        """
+        # If the type is already loaded, returns it.
+        if name in self._loaded_types:
+            return self._loaded_types[name]
+
+        # Otherwise, check if the type dialect is registered.
+        dialect_name, _ = Dialect.split_name(name)
+        if (dialect_name in self._registered_dialects) and (
+            dialect_name not in self._loaded_dialects
+        ):
+            self.load_registered_dialect(dialect_name)
+            return self.get_optional_type(name)
+
+        # If the dialect is unregistered, but the context allows unregistered
+        # dialects, return an UnregisteredAttr.
+        if self.allow_unregistered:
+            from xdsl.dialects.builtin import UnregisteredAttr
+
+            attr_type = UnregisteredAttr.with_name_and_type(name, True)
+            self._loaded_types[name] = attr_type
+            return attr_type
+
+    def get_type(self, name: str) -> type[Attribute]:
+        """
+        Get a type definition from its name.
+        If the type is not registered, raise an exception unless unregistered
+        types are allowed in the context, in which case return an UnregisteredAttr.
+        """
+        if attr_type := self.get_optional_type(name):
+            return attr_type
+        raise Exception(f"Type {name} is not registered")
```

Comment:
KeyError?

Reply:
I added a new exception for the unregistered things (dialects, ops, types, attributes)